block-layer/codecs/dag-pb.md

```diff
@@ -36,50 +36,58 @@ message PBNode {
 The objects link names are specified in the 'Name' field of the PBLink object.
 All link names in an object must either be omitted or unique within the object.
 
+Many Protobuf libraries will ignore (either silently or by some other mechanism to capture missing data) "unrecognized" data within a PB block. Where possible, **DAG-PB decoders should reject blocks with data that doesn't not contribute to the PB schema above**. Additional data that does not appear in the Data Model form of the block should result in the rejection of the block as bad.
+
 ## Logical Format
 
 When we handle DAG-PB content at the Data Model level, we treat these objects as maps.
 
 This layout can be expressed with [IPLD Schemas](../../schemas/README.md) as:
 
 ```ipldsch
-type PBLink struct {
-  Hash Bytes
-  Name optional String
-  Tsize optional Int
-}
-
 type PBNode struct {
   Links optional [PBLink]
```

Comment:
If I understood correctly that the protobuf wire format doesn't disambiguate between a lack of a list here vs a zero length list... maybe it would be best to describe it in our schemas as `Links [PBLink]` _without_ the optional modifer on the field?

If we have the `optional` modifier on the field, that means it _adds_ `Absent` to the cardinality (it +1's the cardinality), but we really have no way of *taking away* the empty list as a valid value (if we could, that'd be -1 on the cardinality, and balance things -- but we don't have such an ability).

Just having the codec handle the translation from {empty list}<->{absense of encoded bytes} seems correct.

Reply:
Yeah, so I've been wanting to lean this way but the thing that's stopped me is that it creates a lot of situations where you're forcing the use (and allocation) of an empty list where you don't want anything. For decoding a dag-pb block with just data in it you're going to need to throw in an extra array just to satisfy this case. When creating an object manually in our APIs you're going to have to allocate an array. In JavaScript this means that I have to force users to give me `{ Data: [...], Links: [] }` when they're just wanting to ingest terminal data blocks. This isn't a huge inconvenience as I've created this `prepare()` method to help get the forms right before instantiating, but it adds a level of hostility that I'm trying to avoid. I think in ipld-prime you'd have data structures that have an empty array for `Links` when you create a new `PBNode` anyway so I suppose you get around this hostility fairly easily.

But.. I take your point on cardinality. And the reason why I've been tempted in this direction is that there's something to be said on the other end of this API where you _always_ get an array in that position and don't need to check for its existence. If I can write a traversal that safely assumes `Links` is always there then it's much nicer than the incessant checking.

I'm going to ponder this today and maybe try out some code.